pkg/services/notary/notary.go

```diff
@@ -0,0 +1,376 @@
+	// nSigs <= nKeys for multisignature request.
+	// nSigs is 0 when all received requests were invalid, so check request.typ before access to nSigs.
+	nSigs uint8
+	// nSigsCollected is the number of already collected signatures
+	nSigsCollected uint8
+
+	// sigs is a map of partial multisig invocation scripts [opcode.PUSHDATA1+64+signatureBytes] grouped by public keys
+	sigs map[*keys.PublicKey][]byte
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]*request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, nSigs, pubs, validationErr := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, exists := n.requests[payload.MainTransaction.Hash()]
+	if exists {
+		for _, fb := range r.fallbacks {
+			if fb.Hash().Equals(payload.FallbackTransaction.Hash()) {
+				return // then we already have processed this request
+			}
+		}
+		if nvbFallback < r.minNotValidBefore {
+			r.minNotValidBefore = nvbFallback
+		}
+		if r.typ == Unknown && validationErr == nil {
+			r.typ = typ
+			r.nSigs = nSigs
+		}
+		if r.typ != Unknown && r.nSigsCollected >= r.nSigs { // already collected sufficient number of signatures to complete main transaction
+			r.fallbacks = append(r.fallbacks, payload.FallbackTransaction) // but we still accept extra fallbacks without adding new signatures to main tx
+			return
+		}
+	} else {
+		r = &request{
+			nSigs:             nSigs,
+			main:              payload.MainTransaction,
+			typ:               typ,
+			minNotValidBefore: nvbFallback,
+		}
+		n.requests[payload.MainTransaction.Hash()] = r
+	}
+	r.fallbacks = append(r.fallbacks, payload.FallbackTransaction)
+	if validationErr == nil {
+	loop:
+		for i, w := range payload.MainTransaction.Scripts {
+			if payload.MainTransaction.Signers[i].Account.Equals(n.Config.Chain.GetNotaryContractScriptHash()) {
+				continue
+			}
+			if len(w.VerificationScript) != 0 {
```

Comment:
`InvocationScript`? We're effectively collecting invocation scripts for `Signature`.

Reply:
Fixed.